cmd/lotus-shed/indexes.go

```diff
@@ -387,6 +392,192 @@ var backfillEventsCmd = &cli.Command{
+			return err
+		}
+		if cctx.IsSet("from") {
+			// we need to fetch the tipset after the epoch being specified since we will need to advance currTs
+			currTs, err = api.ChainGetTipSetAfterHeight(ctx, abi.ChainEpoch(cctx.Int("from")+1), currTs.Key())
+			if err != nil {
+				return err
+			}
+		}
+
+		logGood := cctx.Bool("log-good")
+
+		// advance currTs by one epoch and maintain prevTs as the previous tipset (this allows us to easily use the ChainGetParentMessages/Receipt API)
+		prevTs := currTs
+		currTs, err = api.ChainGetTipSet(ctx, currTs.Parents())
+		if err != nil {
+			return fmt.Errorf("failed to load tipset %s: %w", prevTs.Parents(), err)
+		}
+
+		epochs := cctx.Int("epochs")
+		if epochs <= 0 {
+			epochs = math.MaxInt32
+		}
+
+		basePath, err := homedir.Expand(cctx.String("repo"))
+		if err != nil {
+			return err
+		}
+
+		dbPath := path.Join(basePath, "sqlite", "events.db")
+		db, err := sql.Open("sqlite3", dbPath+"?mode=ro")
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			err := db.Close()
+			if err != nil {
+				fmt.Printf("ERROR: closing db: %s", err)
+			}
+		}()
+
+		stmtEventCount, err := db.Prepare(eventCount)
+		if err != nil {
+			return err
+		}
+		stmtEntryCount, err := db.Prepare(entryCount)
+		if err != nil {
+			return err
+		}
+		stmtTipsetSeen, err := db.Prepare(tipsetSeen)
+		if err != nil {
+			return err
+		}
+
+		processHeight := func(ctx context.Context, ts *types.TipSet, receipts []*types.MessageReceipt) error {
+			tsKeyCid, err := ts.Key().Cid()
+			if err != nil {
+				return fmt.Errorf("failed to get tipset key cid: %w", err)
+			}
+
+			var expectEvents int
+			var expectEntries int
+
+			for _, receipt := range receipts {
+				if receipt.ExitCode != exitcode.Ok || receipt.EventsRoot == nil {
+					continue
+				}
+				events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
+				if err != nil {
+					return fmt.Errorf("failed to load events for tipset %s: %w", currTs, err)
+				}
+				expectEvents += len(events)
+				for _, event := range events {
+					expectEntries += len(event.Entries)
+				}
+			}
+
+			var problems []string
+
+			var seenHeight int
+			var seenReverted int
+			if err := stmtTipsetSeen.QueryRow(tsKeyCid.Bytes()).Scan(&seenHeight, &seenReverted); err != nil {
+				if err == sql.ErrNoRows {
+					if expectEvents > 0 {
+						problems = append(problems, "not in events_seen table")
+					} else {
+						problems = append(problems, "zero-event epoch not in events_seen table")
+					}
+				} else {
+					return fmt.Errorf("failed to check if tipset is seen: %w", err)
+				}
+			} else {
+				if seenHeight != int(ts.Height()) {
+					problems = append(problems, fmt.Sprintf("events_seen height mismatch (%d)", seenHeight))
+				}
+				if seenReverted != 0 {
+					problems = append(problems, "events_seen marked as reverted")
+				}
+			}
+
+			var actualEvents int
+			if err := stmtEventCount.QueryRow(tsKeyCid.Bytes()).Scan(&actualEvents); err != nil {
```

Comment:
This is fine because the events Index maps events/entries against their inclusion tipset and not the execution tipset.